File: lab5/Dictionary.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "Dictionary.h"

typedef struct NodeObj{
	char* key;
	char* value;
	struct NodeObj* next;
} NodeObj;

typedef NodeObj* Node;

Node newNode(char* k, char* v);

Node findKey(Dictionary D, char* key);

void freeNode(Node* pN);


Node newNode(char* k, char* v){
	Node N = malloc(sizeof(NodeObj));
	assert(N != NULL);
	N->key = k;
	N->value = v;
	N->next = NULL;
	return N;
}

void freeNode(Node* pN){
	if( pN != NULL && *pN != NULL){
		free(*pN);
		*pN = NULL;
	}
}

typedef struct DictionaryObj{
	Node head;
	int numPairs;
} DictionaryObj;

Dictionary newDictionary(void){
	Dictionary D = malloc(sizeof(DictionaryObj));
	assert(D != NULL);
	D->head = NULL;
	D->numPairs = 0;
	return D;
}

int isEmpty(Dictionary D){
	if(D == NULL){
		fprintf(stderr, "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	return D->numPairs == 0;
}

int size(Dictionary D){
	return D->numPairs;
}

Node findKey(Dictionary D, char* k){
	Node N = D->head;
	while(N !=NULL){
		if(strcmp(k, N->key) == 0) return N;
		N = N->next;
	}
	return NULL;
}

Node findPrev(Dictionary D, char* k){
	Node N = D->head;
	while(N->next != NULL){
		if(N->next->key == k) return N;
		N = N->next;
	}
	freeNode(&N);
	return NULL;

}




char* lookup(Dictionary D, char* k){
	Node N;
	N = findKey(D, k);
	if(N == NULL) return NULL;
	return N->value;
}

void insert(Dictionary D, char* k, char* v){
	Node N;
	if(D == NULL){
		fprintf(stderr, "Dictionary Error: insert() called on NULL Dictionary reference");
		exit(EXIT_FAILURE);
	}
	if(lookup(D, k) != NULL){
		fprintf(stderr, "Dictionary Error: insert() called with existing key");
		exit(EXIT_FAILURE);
		}
	N = newNode(k, v);
	N->next = D->head;
	D->head = N;
	D->numPairs++;
}

void delete(Dictionary D, char* k){
	Node N;
	Node P;
	if(lookup(D, k) == NULL){
		fprintf(stderr, "Dictionary Error: delete() called on non-existing key");
		exit(EXIT_FAILURE);
	}
	N = findKey(D, k);
	if(N == D->head){
		D->head = D->head->next;
		freeNode(&N);
	}
	else{
		N = D->head;
		while(N->next->key != k){
			N = N->next;
			} 
		P = N->next;
		N->next = P->next;
		freeNode(&P);
	}
	D->numPairs--;
}

void deleteAll(Dictionary D){
	Node N = D->head;
	Node P;
	while(N != NULL){
		P = N;
		N = N->next;
		freeNode(&P);
	}
}

void makeEmpty(Dictionary D){
	Node N, M;
	if(D != NULL){
		N = D->head;
		M = N->next;
		freeNode(&N);
		N = M;
		M = N->next;
	}
	D->head = NULL;
	D->numPairs = 0;
}


void freeDictionary(Dictionary* pD){
	if(pD != NULL && *pD != NULL){
		if(!isEmpty(*pD)) makeEmpty(*pD);
		free(*pD);
		*pD = NULL; 
	}
}

void print_recurse(Node N, FILE* out){
	if(N!=NULL){
		print_recurse(N->next, out);
		fprintf(out, "%s %s\n", N->key, N->value);
		}
	}

void printDictionary(FILE* out, Dictionary D){
	if(D == NULL){
		fprintf(stderr, "Dictionary Error: printDictionary() called on NULL Dictionary reference");
		exit(EXIT_FAILURE);
		}
	if(D->numPairs == 0){
		fprintf(stderr, "Dictionary Error: printDictionary() called on empty Dictionary");
		exit(EXIT_FAILURE);
		}
 	print_recurse(D->head, out);
	
}

File: pa3/Dictionary.java
public class Dictionary implements DictionaryInterface {
	private class Node {
		String key;
		String value;
		Node next;

		Node(String k, String val){
			key = k;
			value = val;
			next = null;
		}
	}

	private Node head;
	private int numPairs;

	private Node findKey(String k){
		Node N = head;
		while(N != null){
			if(N.key == k) return N;
			N = N.next;
		}
		return null;
	}

	private Node findPrevNode(String k){
		Node P = head;
		Node C = head.next;
		while(C != null){
			if(C.key == k) return P;
			C = C.next;
			P = P.next;
		}
		return null;
	}

	public Dictionary(){
		head = null;
		numPairs = 0;
	}

	public boolean isEmpty(){
		return(numPairs == 0);
	}

	public int size(){
		return(numPairs);
	}

	public void insert(String key, String value) throws DuplicateKeyException{
		if(lookup(key) != null){
			throw new DuplicateKeyException("Dictionary Error: insert() called with existing key");
		}
		Node N = new Node(key, value);
		N.next = head;
		head = N;
		numPairs++;
	}


	public String lookup(String k){
		Node A = findKey(k);
		if(A == null) return null;
		return A.value;
	}

	public void delete(String key) throws KeyNotFoundException{
		if(lookup(key) == null){
			throw new KeyNotFoundException("Dictionary Error: delete() called with non-existing key");
		}
		Node A = findKey(key);
		Node Pre = head;
		if(Pre == A){
			head = head.next;
			A = null;
			numPairs--;
		}
		else{
			for( ; Pre.next != A; Pre = Pre.next){};
			Pre.next = A.next;
			A.next = null;
			numPairs--;
		}
	}


	public void makeEmpty(){
		head = null;
		numPairs = 0;
	}

	public String toString(){
		Node N = head;
		StringBuffer sb = new StringBuffer();
		sb = reversePairs(N, sb);
		return new String(sb);
	}

	public StringBuffer reversePairs(Node N, StringBuffer sb){
		if(N != null){
			reversePairs(N.next, sb);
			sb.append(N.key).append(" ").append(N.value).append("\n");
		}
		return sb;
	}	

}
